qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -259,32 +278,50 @@ def run(self, dag):
             # After all free gates are exhausted, heuristically find
             # the best swap and insert it. When two or more swaps tie
             # for best score, pick one randomly.
+
             if extended_set is None:
                 extended_set = self._obtain_extended_set(dag, front_layer)
-            swap_scores = {}
-            for swap_qubits in self._obtain_swaps(front_layer, current_layout):
-                trial_layout = current_layout.copy()
-                trial_layout.swap(*swap_qubits)
-                score = self._score_heuristic(
-                    self.heuristic, front_layer, extended_set, trial_layout, swap_qubits
+            extended_set_list = EdgeList(len(extended_set))
+            for x in extended_set:
+                extended_set_list.append(
+                    self._bit_indices[x.qargs[0]], self._bit_indices[x.qargs[1]]
+                )
+
+            front_layer_list = EdgeList(len(front_layer))
+            for x in front_layer:
+                front_layer_list.append(
+                    self._bit_indices[x.qargs[0]], self._bit_indices[x.qargs[1]]
                 )
-                swap_scores[swap_qubits] = score
-            min_score = min(swap_scores.values())
-            best_swaps = [k for k, v in swap_scores.items() if v == min_score]
-            best_swaps.sort(key=lambda x: (self._bit_indices[x[0]], self._bit_indices[x[1]]))
+            swap_candidates = self._obtain_swaps(front_layer, layout, canonical_register)
+            swap_candidates = [
+                (self._bit_indices[x[0]], self._bit_indices[x[1]]) for x in swap_candidates
+            ]
+            swap_scores = SwapScores(swap_candidates)
+
+            best_swaps = sabre_score_heuristic(
+                front_layer_list,
+                layout,
+                swap_scores,
+                extended_set_list,
+                self.dist_matrix,
+                self.qubits_decay,
+                self.heuristic,
+            )
             best_swap = rng.choice(best_swaps)
+            best_swap_qargs = [canonical_register[x] for x in best_swap]
             swap_node = self._apply_gate(
                 mapped_dag,
-                DAGOpNode(op=SwapGate(), qargs=best_swap),
-                current_layout,
+                DAGOpNode(op=SwapGate(), qargs=best_swap_qargs),
+                layout,
                 canonical_register,
             )
-            current_layout.swap(*best_swap)
+            layout.swap_logic(*best_swap)
             ops_since_progress.append(swap_node)
 
             num_search_steps += 1
-            if num_search_steps % DECAY_RESET_INTERVAL == 0:
-                self._reset_qubits_decay()
+            if num_search_steps >= DECAY_RESET_INTERVAL == 0:
```

Comment:
I suspect this never triggers because of the coupled condition `DECAY_RESET_INTERVAL == 0` (and it never does).

Reply:
Heh, yeah I meant to do this as >= interval. A big typo/oversight that I forgot to drop the == 0. I'll just revert this to what it was doing before

Reply:
Fixed in: https://github.com/Qiskit/qiskit-terra/pull/7977/commits/b76284e351fed7ea0e3ee23b5b4501ee82e0214a